src/services/completions.ts

```diff
@@ -2197,4 +2198,28 @@ namespace ts.Completions {
     function hasIndexSignature(type: Type): boolean {
         return !!type.getStringIndexType() || !!type.getNumberIndexType();
     }
+
+    function isInvalidTrigger(sourceFile: SourceFile, triggerCharacter: string | undefined, contextToken: Node, position: number): boolean {
+        switch (triggerCharacter) {
+            case '"':
+            case "'":
+            case "`":
+                switch (contextToken.kind) {
+                    case SyntaxKind.StringLiteral:
+                    case SyntaxKind.NoSubstitutionTemplateLiteral:
+                    case SyntaxKind.TemplateExpression:
+                    case SyntaxKind.TaggedTemplateExpression:
+                        // Don't automatically bring up completions at closing quote
+                        return position === contextToken.parent.end;
+                    default:
+                        if (!isInComment(sourceFile, position, contextToken)) Debug.failBadSyntaxKind(contextToken);

```

Comment:
not sure you need this. We are doing some extra work just to know it is invalid. i would just return directly and avoid walking the tree every time in a comment just for an assert.